stdlib/LinearAlgebra/src/triangular.jl

```diff
@@ -945,7 +945,7 @@ function _trilkron!(C, A, B)
                     C[inB+k, jnB+l] = Aij * B[k, l]
                 end
                 for k = (l+1):n_B
-                    C[inB+l, jnB+k] = zero(eltype(C))
+                    C[inB+l, jnB+k] = zero(C[1])
```

Comment:
I'm not sure if this will necessarily be correct (e.g. if `C[1]` has a different type from other elements). There was a similar discussion in https://github.com/JuliaLang/julia/pull/55630. I wonder if we may compute the zero element here from the input matrices? Or will that make this considerably slower?

Reply:
How so? I did this for triangular matrices precisely because unlike `Symmetric`/`Hermitian` the types of the elements of `C` are always the same.

Reply:
The elements may be different-sized matrices

Reply:
Here is the first example where the results change:
```julia
julia> U = UpperTriangular(Matrix{Union{Missing,Int}}(fill(2,2,2))); U[1,1] = missing;

julia> U
2×2 UpperTriangular{Union{Missing, Int64}, Matrix{Union{Missing, Int64}}}:
  missing  2
 ⋅         2

julia> kron(U, U)
4×4 Matrix{Union{Missing, Int64}}:
  missing   missing   missing  4
  missing   missing  0         4
  missing  0          missing  4
 0         0         0         4

julia> kron(U, U)[2,3]
0
```
vs this PR
```julia
julia> kron(U, U)
4×4 UpperTriangular{Union{Missing, Int64}, Matrix{Union{Missing, Int64}}}:
  missing   missing   missing  4
     ⋅      missing   missing  4
     ⋅         ⋅      missing  4
 ⋅         ⋅         ⋅         4

julia> kron(U, U)[2,3]
missing
```

I understand that the next one is a somewhat contrived example, but this is also a case where the results change:
```julia
julia> U = UpperTriangular(reshape([ones(2,1), zeros(1,1), ones(2,1), ones(1,1)], 2, 2))
2×2 UpperTriangular{Matrix{Float64}, Matrix{Matrix{Float64}}}:
 [1.0; 1.0;;]  [1.0; 1.0;;]
      ⋅        [1.0;;]

julia> U2 = UpperTriangular(reshape([ones(1,2), zeros(1,2), ones(1,1), fill(4,1,1)], 2, 2))
2×2 UpperTriangular{Matrix{Float64}, Matrix{Matrix{Float64}}}:
 [1.0 1.0]  [1.0;;]
    ⋅       [4.0;;]

julia> kron(U, U2)
4×4 Matrix{Matrix{Float64}}:
 [1.0 1.0; 1.0 1.0]  [1.0; 1.0;;]  [1.0 1.0; 1.0 1.0]  [1.0; 1.0;;]
 [0.0; 0.0;;]        [4.0; 4.0;;]  [0.0; 0.0;;]        [4.0; 4.0;;]
 [0.0 0.0; 0.0 0.0]  [0.0; 0.0;;]  [1.0 1.0]           [1.0;;]
 [0.0; 0.0;;]        [0.0; 0.0;;]  [0.0;;]             [4.0;;]

julia> kron(U, U2)[2,3]
2×1 Matrix{Float64}:
 0.0
 0.0
```
vs this PR
```julia
julia> kron(U, U2)
4×4 UpperTriangular{Matrix{Float64}, Matrix{Matrix{Float64}}}:
 [1.0 1.0; 1.0 1.0]  [1.0; 1.0;;]        [1.0 1.0; 1.0 1.0]  [1.0; 1.0;;]
      ⋅              [4.0; 4.0;;]        [0.0 0.0; 0.0 0.0]  [4.0; 4.0;;]
         ⋅                   ⋅           [1.0 1.0]           [1.0;;]
      ⋅                   ⋅                 ⋅                [4.0;;]

julia> kron(U, U2)[2,3]
2×2 Matrix{Float64}:
 0.0  0.0
 0.0  0.0
```

Reply:
fair enough, I pushed a change, now it has the behaviour you want:
```julia
julia> using LinearAlgebra

julia> U = UpperTriangular(Matrix{Union{Missing,Int}}(fill(2,2,2))); U[1,1] = missing;

julia> kron(U, U)
4×4 UpperTriangular{Union{Missing, Int64}, Matrix{Union{Missing, Int64}}}:
  missing   missing   missing  4
     ⋅      missing  0         4
     ⋅     ⋅          missing  4
 ⋅         ⋅         ⋅         4

julia> U = UpperTriangular(reshape([ones(2,1), zeros(1,1), ones(2,1), ones(1,1)], 2, 2))
2×2 UpperTriangular{Matrix{Float64}, Matrix{Matrix{Float64}}}:
 [1.0; 1.0;;]  [1.0; 1.0;;]
      ⋅        [1.0;;]

julia> U2 = UpperTriangular(reshape([ones(1,2), zeros(1,2), ones(1,1), fill(4,1,1)], 2, 2))
2×2 UpperTriangular{Matrix{Float64}, Matrix{Matrix{Float64}}}:
 [1.0 1.0]  [1.0;;]
    ⋅       [4.0;;]

julia> kron(U, U2)
4×4 UpperTriangular{Matrix{Float64}, Matrix{Matrix{Float64}}}:
 [1.0 1.0; 1.0 1.0]  [1.0; 1.0;;]  [1.0 1.0; 1.0 1.0]  [1.0; 1.0;;]
      ⋅              [4.0; 4.0;;]  [0.0; 0.0;;]        [4.0; 4.0;;]
         ⋅                ⋅        [1.0 1.0]           [1.0;;]
      ⋅                   ⋅           ⋅                [4.0;;]
```
Note that the following behaviour might still be unexpected:
```julia
julia> a = [1 1;0 1]; b = [1 missing;0 1];

julia> kron(a,b)
4×4 Matrix{Union{Missing, Int64}}:
 1   missing  1   missing
 0  1         0  1
 0   missing  1   missing
 0  0         0  1

julia> kron(UpperTriangular(a),UpperTriangular(b))
4×4 UpperTriangular{Union{Missing, Int64}, Matrix{Union{Missing, Int64}}}:
 1          missing  1          missing
     ⋅     1          missing  1
 ⋅             ⋅     1          missing
     ⋅     ⋅             ⋅     1
```
But that's unavoidable because the `0` in the lower triangular of `b` becomes a `missing` when we wrap it with `UpperTriangular`.

Reply:
Thanks for the change! Could you also add a test, perhaps for the `missing` case?